File: 2026/07-July/01-Wednesday/solution.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var maximumSafenessFactor = function(grid) {
    const n = grid.length;
    
    // If start or end contains a thief, return 0
    if (grid[0][0] === 1 || grid[n - 1][n - 1] === 1) {
        return 0;
    }

    // Step 1: Multi-source BFS
    const distToThief = Array.from({ length: n }, () => Array(n).fill(-1));
    const q = [];
    let head = 0; // Using head pointer for O(1) queue shift operations

    for (let r = 0; r < n; r++) {
        for (let c = 0; c < n; c++) {
            if (grid[r][c] === 1) {
                q.push([r, c]);
                distToThief[r][c] = 0;
            }
        }
    }

    const directions = [[0, 1], [1, 0], [0, -1], [-1, 0]];

    while (head < q.length) {
        const [r, c] = q[head++];
        
        for (const [dr, dc] of directions) {
            const nr = r + dr;
            const nc = c + dc;
            
            if (nr >= 0 && nr < n && nc >= 0 && nc < n && distToThief[nr][nc] === -1) {
                distToThief[nr][nc] = distToThief[r][c] + 1;
                q.push([nr, nc]);
            }
        }
    }

    // Step 2: Modified Dijkstra using LeetCode's built-in MaxPriorityQueue
    const maxHeap = new MaxPriorityQueue({ priority: (x) => x.safeness });
    const visited = Array.from({ length: n }, () => Array(n).fill(false));

    maxHeap.enqueue({ safeness: distToThief[0][0], r: 0, c: 0 });
    visited[0][0] = true;

    while (!maxHeap.isEmpty()) {
        const { safeness: currentSafeness, r, c } = maxHeap.dequeue().element;

        // Reached destination
        if (r === n - 1 && c === n - 1) {
            return currentSafeness;
        }

        for (const [dr, dc] of directions) {
            const nr = r + dr;
            const nc = c + dc;
            
            if (nr >= 0 && nr < n && nc >= 0 && nc < n && !visited[nr][nc]) {
                visited[nr][nc] = true;
                const nextSafeness = Math.min(currentSafeness, distToThief[nr][nc]);
                maxHeap.enqueue({ safeness: nextSafeness, r: nr, c: nc });
            }
        }
    }

    return 0;
};

File: 2026/07-July/01-Wednesday/solution.java
import java.util.*;

class Solution {
    public int maximumSafenessFactor(List<List<Integer>> grid) {
        int n = grid.size();
        
        // If start or end contains a thief, return 0
        if (grid.get(0).get(0) == 1 || grid.get(n - 1).get(n - 1) == 1) {
            return 0;
        }
        
        int[][] distToThief = new int[n][n];
        for (int[] row : distToThief) {
            Arrays.fill(row, -1);
        }
        
        Queue<int[]> q = new LinkedList<>();
        
        // Step 1: Initialize Multi-source BFS
        for (int r = 0; r < n; r++) {
            for (int c = 0; c < n; c++) {
                if (grid.get(r).get(c) == 1) {
                    q.offer(new int[]{r, c});
                    distToThief[r][c] = 0;
                }
            }
        }
        
        int[][] directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        
        while (!q.isEmpty()) {
            int[] curr = q.poll();
            int r = curr[0];
            int c = curr[1];
            
            for (int[] dir : directions) {
                int nr = r + dir[0];
                int nc = c + dir[1];
                
                if (nr >= 0 && nr < n && nc >= 0 && nc < n && distToThief[nr][nc] == -1) {
                    distToThief[nr][nc] = distToThief[r][c] + 1;
                    q.offer(new int[]{nr, nc});
                }
            }
        }
        
        // Step 2: Modified Dijkstra to find max safeness path
        // PriorityQueue ordered by safeness factor descending: (safeness, r, c)
        PriorityQueue<int[]> maxHeap = new PriorityQueue<>((a, b) -> b[0] - a[0]);
        boolean[][] visited = new boolean[n][n];
        
        maxHeap.offer(new int[]{distToThief[0][0], 0, 0});
        visited[0][0] = true;
        
        while (!maxHeap.isEmpty()) {
            int[] curr = maxHeap.poll();
            int currentSafeness = curr[0];
            int r = curr[1];
            int c = curr[2];
            
            if (r == n - 1 && c == n - 1) {
                return currentSafeness;
            }
            
            for (int[] dir : directions) {
                int nr = r + dir[0];
                int nc = c + dir[1];
                
                if (nr >= 0 && nr < n && nc >= 0 && nc < n && !visited[nr][nc]) {
                    visited[nr][nc] = true;
                    int nextSafeness = Math.min(currentSafeness, distToThief[nr][nc]);
                    maxHeap.offer(new int[]{nextSafeness, nr, nc});
                }
            }
        }
        
        return 0;
    }
}
